main.lua

```diff
@@ -957,9 +959,52 @@ function onCursorRight(bufpane)    clearAutocomplete() end
 function onCursorStart(bufpane)    clearAutocomplete() end
 function onCursorEnd(bufpane)      clearAutocomplete() end
 
-function onUndo(bp) fullyUpdate(bp.Buf) end
-function onRedo(bp) fullyUpdate(bp.Buf) end
+function preUndo(bp)
+    undoStackLengthBefore = bp.Buf.UndoStack:Len()
+end
+function onUndo(bp)
+    local numUndos = undoStackLengthBefore - bp.Buf.UndoStack:Len()
+    return handleUndosRedos(bp.Buf, bp.Buf.RedoStack.Top, numUndos)
+end
+
+function preRedo(bp)
+    undoStackLengthBefore = bp.Buf.UndoStack:Len()
+end
+function onRedo(bp)
+    local numRedos = bp.Buf.UndoStack:Len() - undoStackLengthBefore
+    return handleUndosRedos(bp.Buf, bp.Buf.UndoStack.Top, numRedos)
+end
+
+function handleUndosRedos(buf, elem, numUndos)
+    if next(activeConnections) == nil then return end
+
+    local TEXT_EVENT = {INSERT = 1, REMOVE = -1, REPLACE = 0}
+    local tevents = {}
+    for i = 1, numUndos do
+        table.insert(tevents, elem.Value)
+        elem = elem.Next
+    end
+
+    local changes = {}
+    for i = 1, #tevents do
+        local tev = tevents[#tevents + 1 - i]
+        for _, delta in userdataIterator(tev.Deltas) do
+            local text = ""
+            local range = LSPRange.fromDelta(delta)
+
+            if tev.EventType == TEXT_EVENT.INSERT then
```

Comment:
And this probably should change depending of the stack that you are (you can make another "enumeration" for that or make TEXT_EVENT global again an pass it to `handleUndosRedos()`).

Probably you should watch what is in the Stack and how it changes. 

```lua
-- comment your `log` function to reduce noise
local function dumpEventStack(name, stack)
    micro.Log("--------------------- "..name.." ---------------------")
    micro.Log("Stack", stack)
    if stack then
        local elem = stack.Top
        for _=0, stack:Len() do
            if not elem then break end
            micro.Log(name.."::elem.Value", elem.Value)
            elem = elem.Next
        end
    end
    micro.Log("------------------------------------------")
end
```

If I remember correctly when a TextEvent is passed from one stack to another the typeEvent is inverted (1 to -1, and -1 to 1. INSERT to REMOVE and REMOVE to INSERT), so I think it should contain the same content with ONLY type inverted. Now you have to check in which stacks are you, and in base of that change the TEXT_EVENT.INSERT/REMOVE condition.

Reply:
> And this probably should change depending of the stack that you are (you can make another "enumeration" for that or make TEXT_EVENT global again an pass it to `handleUndosRedos()`).

Are you sure? I thought so too initially but it seemed to keep all diagnostics correctly in sync no matter what I tried to do with undo/redo without doing that – so I think micro itself must be inverting the tevents before pushing them onto the stack.

Reply:
Probably in what I had in mind yes, in yours maybe not? I based all the comments in a wrong premise.